core/src/main/scala/kafka/metrics/clientmetrics/CmClientInformation.scala

```diff
@@ -0,0 +1,104 @@
+package kafka.metrics.clientmetrics
+
+import kafka.Kafka.info
+import kafka.metrics.clientmetrics.CmClientInformation._
+import kafka.network.RequestChannel
+import org.apache.kafka.common.errors.InvalidConfigurationException
+
+import scala.collection.mutable
+
+
+/**
+ * Information from the client's metadata is gathered from the client's request.
+ */
+object CmClientInformation {
+  val CLIENT_ID = "client_id"
+  val CLIENT_INSTANCE_ID = "client_instance_id"
+  val CLIENT_SOFTWARE_NAME = "client_software_name"
+  val CLIENT_SOFTWARE_VERSION = "client_software_version"
+  val CLIENT_SOURCE_ADDRESS = "client_source_address"
+  val CLIENT_SOURCE_PORT = "client_source_port"
+
+  val matchersList = List(CLIENT_ID, CLIENT_INSTANCE_ID, CLIENT_SOFTWARE_NAME,
+                          CLIENT_SOFTWARE_VERSION, CLIENT_SOURCE_ADDRESS, CLIENT_SOURCE_PORT)
+
+  def getMatchers = matchersList
+
+  def apply(request: RequestChannel.Request, clientInstanceId: String): CmClientInformation = {
+    val instance = new CmClientInformation
+    instance.init(clientInstanceId, request.context.clientId(), request.context.clientInformation.softwareName(),
+                  request.context.clientInformation.softwareVersion(), request.context.clientAddress.getHostAddress,
+                  request.context.clientAddress.getHostAddress )
+    instance
+  }
+
+  def apply(clientInstanceId: String, clientId: String, softwareName: String,
+            softwareVersion: String, hostAddress: String, port: String): CmClientInformation = {
+    val instance = new CmClientInformation
+    instance.init(clientInstanceId, clientId, softwareName, softwareVersion, hostAddress, port )
+    instance
+  }
+
+  /**
+   * Parses the client matching patterns and builds a map with entries that has
+   * (PatternName, PatternValue) as the entries.
+   *  Ex: "VERSION=1.2.3" would be converted to a map entry of (Version, 1.2.3)
+   * @param patterns List of client matching pattern strings
+   * @return
+   */
+  def parseMatchingPatterns(patterns: List[String]) : Map[String, String] = {
+    val patternsMap = mutable.Map[String, String]()
+    patterns.foreach(x =>  {
+      val nameValuePair = x.split("=").map(x => x.trim)
+      if (nameValuePair.size != 2 || !matchersList.contains(nameValuePair(0))) {
```

Comment:
It is not impossible for a client to report a client_software_version that contains a "=",  and this here will not be able to match that. So perhaps of requiring exactly two arguments we just split the string once, on the first = occurence, and let everything after be left verbatim in place?

Reply:
I am good with it, I will change the parser code to account that.